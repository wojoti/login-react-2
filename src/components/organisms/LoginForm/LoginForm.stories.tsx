/* eslint-disable react/function-component-definition */
/* eslint-disable react/jsx-props-no-spreading */
/* eslint-disable import/no-extraneous-dependencies */
import { ComponentMeta, ComponentStory } from "@storybook/react";

import LoginForm from "./LoginForm";

// More on default export: https://storybook.js.org/docs/react/writing-stories/introduction#default-export
export default {
  title: "Organisms/LoginForm",
  component: LoginForm,
} as ComponentMeta<typeof LoginForm>;

// More on component templates: https://storybook.js.org/docs/react/writing-stories/introduction#using-args
const Template: ComponentStory<typeof LoginForm> = (args) => (
  <LoginForm {...args} />
);

export const Default = Template.bind({});
Default.args = {};
