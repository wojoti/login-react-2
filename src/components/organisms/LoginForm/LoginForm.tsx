import { SyntheticEvent, useState } from 'react'
import Button from '../../atoms/Button/Button'
import Checkbox from '../../atoms/Checkbox/Checkbox'
import Column from '../../atoms/Column/Column'
import Label from '../../atoms/Label/Label'
import Link from '../../atoms/Link/Link'
import Row from '../../atoms/Row/Row'
import TextInput from '../../atoms/TextInput/TextInput'

const LoginForm = () => {
  const [email, setEmail] = useState('')
  const [password, setPassword] = useState('')
  const [remember, setRemember] = useState(false)

  const onEmailChange = (event: SyntheticEvent<HTMLInputElement>) => {
    setEmail(event.currentTarget.value)
  }

  const onPasswordChange = (event: SyntheticEvent<HTMLInputElement>) => {
    setPassword(event.currentTarget.value)
  }

  const onRememberChange = (event: SyntheticEvent<HTMLInputElement>) => {
    setRemember(event.currentTarget.checked)
  }

  const onLoginSubmit = (event: SyntheticEvent<HTMLInputElement>) => {
    // event.preventDefault(); //dont use preventDefault when not using form
    const completeAction = {
      email,
      password,
      remember,
      action: event.currentTarget,
    }
    console.log(completeAction)
  }
  return (
    <>
      <Column mt={15}>
        <Label>Email</Label>
        <TextInput type='email' name='email' id='email' onFieldChange={onEmailChange}></TextInput>
      </Column>
      <Column mt={15}>
        <Label>Password</Label>
        <TextInput
          type='password'
          name='password'
          id='password'
          onFieldChange={onPasswordChange}
        ></TextInput>
      </Column>
      <Row mt={10} items='center'>
        <Checkbox name='remember' id='remember' onChange={onRememberChange}></Checkbox>
        <Label>Remember me?</Label>
      </Row>
      <Column mt={15}>
        <Button id='login-submit' name='login-submit' onClick={onLoginSubmit}>
          LOGIN
        </Button>
      </Column>

      <Column items='flex-end'>
        <Link decoration='none' href='/' color='rgb(107 114 128)'>
          Forgot Password?
        </Link>
      </Column>
    </>
  )
}

export default LoginForm
