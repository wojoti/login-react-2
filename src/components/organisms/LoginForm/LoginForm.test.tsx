import { fireEvent, render, screen, within } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import TestRenderer from "react-test-renderer";
import LoginForm, { LoginFormProps } from "./LoginForm";

const props: LoginFormProps = {
  onLoginSubmit: () => {},
  onLinkClick: () => {},
  testId: "test-loginform-id",
};

test("should match snapshot", () => {
  const tree = TestRenderer.create(<LoginForm {...props} />).toJSON();
  expect(tree).toMatchSnapshot();
});

test("should render loginform", () => {
  render(<LoginForm {...props} />);
  const loginformElement = screen.getByTestId("test-loginform-id");
  expect(loginformElement).toBeInTheDocument();
});

test("should render loginform default disabled button ", () => {
  render(<LoginForm {...props} />);
  const loginformElement = screen.getByTestId("test-loginform-id");
  expect(loginformElement).toBeInTheDocument();
  const loginformButtonElement = screen.getByTestId("login-button-id");
  expect(loginformButtonElement).toBeInTheDocument();
  expect(loginformButtonElement).toBeDisabled();
});

test("should render loginform with enabled button - both inputs filled, email regex correct", () => {
  render(<LoginForm {...props} />);
  const loginformElement = screen.getByTestId("test-loginform-id");
  expect(loginformElement).toBeInTheDocument();
  const loginformLoginElement = screen.getByTestId("loginform-textinput-login");
  expect(loginformLoginElement).toBeInTheDocument();
  const textfieldLoginElement = within(loginformLoginElement).getByTestId(
    "textinput-textfield"
  );
  expect(textfieldLoginElement).toBeInTheDocument();
  userEvent.type(textfieldLoginElement, "test@test.eu");

  const loginformPasswordElement = screen.getByTestId(
    "loginform-textinput-password"
  );
  expect(loginformPasswordElement).toBeInTheDocument();
  const textfieldPasswordElement = within(loginformPasswordElement).getByTestId(
    "textinput-textfield"
  );
  expect(textfieldPasswordElement).toBeInTheDocument();
  userEvent.type(textfieldPasswordElement, "pwd");

  const loginformButtonElement = screen.getByTestId("login-button-id");
  expect(loginformButtonElement).toBeInTheDocument();
  expect(loginformButtonElement).not.toBeDisabled();
});

test("should handle onClick on button", () => {
  const handleClick = jest.fn();
  render(<LoginForm {...props} onLoginSubmit={handleClick} />);
  const loginformElement = screen.getByTestId("test-loginform-id");
  expect(loginformElement).toBeInTheDocument();
  const loginformLoginElement = screen.getByTestId("loginform-textinput-login");
  expect(loginformLoginElement).toBeInTheDocument();
  const textfieldLoginElement = within(loginformLoginElement).getByTestId(
    "textinput-textfield"
  );
  expect(textfieldLoginElement).toBeInTheDocument();
  userEvent.type(textfieldLoginElement, "test@test.eu");

  const loginformPasswordElement = screen.getByTestId(
    "loginform-textinput-password"
  );
  expect(loginformPasswordElement).toBeInTheDocument();
  const textfieldPasswordElement = within(loginformPasswordElement).getByTestId(
    "textinput-textfield"
  );
  expect(textfieldPasswordElement).toBeInTheDocument();
  userEvent.type(textfieldPasswordElement, "pwd");

  const loginformButtonElement = screen.getByTestId("login-button-id");
  expect(loginformButtonElement).toBeInTheDocument();
  expect(loginformButtonElement).not.toBeDisabled();

  fireEvent.click(loginformButtonElement);
  expect(handleClick).toHaveBeenCalledTimes(1);
});

test("should render loginform with disabled button - both inputs filled, email regex incorrect", () => {
  render(<LoginForm {...props} />);
  const loginformElement = screen.getByTestId("test-loginform-id");
  expect(loginformElement).toBeInTheDocument();
  const loginformLoginElement = screen.getByTestId("loginform-textinput-login");
  expect(loginformLoginElement).toBeInTheDocument();
  const textfieldLoginElement = within(loginformLoginElement).getByTestId(
    "textinput-textfield"
  );
  expect(textfieldLoginElement).toBeInTheDocument();
  userEvent.type(textfieldLoginElement, "t");

  const loginformPasswordElement = screen.getByTestId(
    "loginform-textinput-password"
  );
  expect(loginformPasswordElement).toBeInTheDocument();
  const textfieldPasswordElement = within(loginformPasswordElement).getByTestId(
    "textinput-textfield"
  );
  expect(textfieldPasswordElement).toBeInTheDocument();
  userEvent.type(textfieldPasswordElement, "pwd");

  const loginformButtonElement = screen.getByTestId("login-button-id");
  expect(loginformButtonElement).toBeInTheDocument();
  expect(loginformButtonElement).toBeDisabled();
});

test("should render loginform with disabled button - login filled, email regex correct", () => {
  render(<LoginForm {...props} />);
  const loginformElement = screen.getByTestId("test-loginform-id");
  expect(loginformElement).toBeInTheDocument();
  const loginformLoginElement = screen.getByTestId("loginform-textinput-login");
  expect(loginformLoginElement).toBeInTheDocument();
  const textfieldLoginElement = within(loginformLoginElement).getByTestId(
    "textinput-textfield"
  );
  expect(textfieldLoginElement).toBeInTheDocument();
  userEvent.type(textfieldLoginElement, "test@test.eu");

  const loginformButtonElement = screen.getByTestId("login-button-id");
  expect(loginformButtonElement).toBeInTheDocument();
  expect(loginformButtonElement).toBeDisabled();
});

test("should render loginform with disabled button - password filled", () => {
  render(<LoginForm {...props} />);
  const loginformElement = screen.getByTestId("test-loginform-id");
  expect(loginformElement).toBeInTheDocument();
  const loginformPasswordElement = screen.getByTestId(
    "loginform-textinput-password"
  );
  expect(loginformPasswordElement).toBeInTheDocument();
  const textfieldPasswordElement = within(loginformPasswordElement).getByTestId(
    "textinput-textfield"
  );
  expect(textfieldPasswordElement).toBeInTheDocument();
  userEvent.type(textfieldPasswordElement, "pwd");

  const loginformButtonElement = screen.getByTestId("login-button-id");
  expect(loginformButtonElement).toBeInTheDocument();
  expect(loginformButtonElement).toBeDisabled();
});

test("should handle onClick on link", () => {
  const handleClick = jest.fn();
  render(<LoginForm {...props} onLinkClick={handleClick} />);
  const loginformElement = screen.getByTestId("test-loginform-id");
  expect(loginformElement).toBeInTheDocument();

  const loginformLinkElement = screen.getByTestId("loginform-link");
  expect(loginformLinkElement).toBeInTheDocument();

  fireEvent.click(loginformLinkElement);
  expect(handleClick).toHaveBeenCalledTimes(1);
});

test("should handle click on checkbox", () => {
  render(<LoginForm {...props} />);
  const loginformElement = screen.getByTestId("test-loginform-id");
  expect(loginformElement).toBeInTheDocument();

  const loginformCheckboxAreaElement = screen.getByTestId(
    "loginform-CheckboxArea"
  );
  expect(loginformCheckboxAreaElement).toBeInTheDocument();

  const checkboxElement = within(loginformCheckboxAreaElement).getByTestId(
    "checkboxarea-checkbox"
  );
  expect(checkboxElement).not.toBeChecked();
  fireEvent.click(checkboxElement);
  expect(checkboxElement).toBeChecked();
});

test("should render email textfield with red border when invalid (eg. regex)", () => {
  render(<LoginForm {...props} />);
  const loginformElement = screen.getByTestId("test-loginform-id");
  expect(loginformElement).toBeInTheDocument();
  const loginformLoginElement = screen.getByTestId("loginform-textinput-login");
  expect(loginformLoginElement).toBeInTheDocument();
  const textfieldLoginElement = within(loginformLoginElement).getByTestId(
    "textinput-textfield"
  );
  expect(textfieldLoginElement).toBeInTheDocument();
  userEvent.type(textfieldLoginElement, "t");
  expect(textfieldLoginElement).toHaveStyleRule(
    "border",
    "2px solid rgb(244 63 94)"
  );
});

test("should not render password textfield with red border when invalid'", () => {
  render(<LoginForm {...props} />);
  const loginformElement = screen.getByTestId("test-loginform-id");
  expect(loginformElement).toBeInTheDocument();
  const loginformLoginElement = screen.getByTestId(
    "loginform-textinput-password"
  );
  expect(loginformLoginElement).toBeInTheDocument();
  const textfieldLoginElement = within(loginformLoginElement).getByTestId(
    "textinput-textfield"
  );
  expect(textfieldLoginElement).toBeInTheDocument();
  userEvent.type(textfieldLoginElement, "t");
  expect(textfieldLoginElement).not.toHaveStyleRule(
    "border",
    "2px solid rgb(244 63 94)"
  );
});
test("should render checkboxarea with checkbox unchecked by default", () => {
  render(<LoginForm {...props} />);
  const loginformElement = screen.getByTestId("test-loginform-id");
  expect(loginformElement).toBeInTheDocument();

  const checkboxareaElement = screen.getByTestId("loginform-CheckboxArea");
  expect(checkboxareaElement).toBeInTheDocument();
  const checkboxareaCheckboxElement = within(checkboxareaElement).getByTestId(
    "checkboxarea-checkbox"
  );
  expect(checkboxareaCheckboxElement).toBeInTheDocument();
  expect(checkboxareaCheckboxElement).not.toBeChecked();
});

test("should switch checkbox status after click", () => {
  render(<LoginForm {...props} />);
  const loginformElement = screen.getByTestId("test-loginform-id");
  expect(loginformElement).toBeInTheDocument();

  const checkboxareaElement = screen.getByTestId("loginform-CheckboxArea");
  expect(checkboxareaElement).toBeInTheDocument();
  const checkboxareaCheckboxElement = within(checkboxareaElement).getByTestId(
    "checkboxarea-checkbox"
  );
  expect(checkboxareaCheckboxElement).toBeInTheDocument();
  fireEvent.click(checkboxareaCheckboxElement);
  expect(checkboxareaCheckboxElement).toBeChecked();
});
