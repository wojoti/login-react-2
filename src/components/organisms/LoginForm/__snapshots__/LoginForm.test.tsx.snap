// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
.c8 {
  box-sizing: border-box;
  margin: 0;
  padding: 0;
  font-family: -apple-system,BlinkMacSystemFont,"Segoe UI","Roboto","Oxygen", "Ubuntu","Cantarell","Fira Sans","Droid Sans","Helvetica Neue", sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  width: 100%;
  height: 32px;
  border-radius: 6px;
  border: 0px solid black;
  background-color: rgb(244 63 94);
  background-color: rgb(244 63 94);
  color: white;
  cursor: pointer;
  box-shadow: 0 25px 50px -12px rgb(0 0 0 / 0.25);
}

.c7 {
  box-sizing: border-box;
  margin: 0;
  padding: 0;
  font-family: -apple-system,BlinkMacSystemFont,"Segoe UI","Roboto","Oxygen", "Ubuntu","Cantarell","Fira Sans","Droid Sans","Helvetica Neue", sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  margin-right: 10px;
  accent-color: #f43f5e;
  width: 1rem;
  height: 1rem;
}

.c1 {
  box-sizing: border-box;
  font-family: -apple-system,BlinkMacSystemFont,"Segoe UI","Roboto","Oxygen", "Ubuntu","Cantarell","Fira Sans","Droid Sans","Helvetica Neue", sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  margin-top: 15px;
  margin-bottom: 0px;
  margin-left: 0px;
  margin-right: 0px;
  padding-top: 0px;
  padding-bottom: 0px;
  padding-left: 0px;
  padding-right: 0px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
}

.c9 {
  box-sizing: border-box;
  font-family: -apple-system,BlinkMacSystemFont,"Segoe UI","Roboto","Oxygen", "Ubuntu","Cantarell","Fira Sans","Droid Sans","Helvetica Neue", sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  margin-top: 0px;
  margin-bottom: 0px;
  margin-left: 0px;
  margin-right: 0px;
  padding-top: 0px;
  padding-bottom: 0px;
  padding-left: 0px;
  padding-right: 0px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
}

.c3 {
  box-sizing: border-box;
  margin: 0;
  padding: 0;
  font-family: -apple-system,BlinkMacSystemFont,"Segoe UI","Roboto","Oxygen", "Ubuntu","Cantarell","Fira Sans","Droid Sans","Helvetica Neue", sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  color: inherit;
  font-size: 14px;
}

.c10 {
  box-sizing: border-box;
  margin: 0;
  padding: 0;
  font-family: -apple-system,BlinkMacSystemFont,"Segoe UI","Roboto","Oxygen", "Ubuntu","Cantarell","Fira Sans","Droid Sans","Helvetica Neue", sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: rgb(107 114 128);
  font-size: 14px;
  background: none !important;
  border: none;
  padding: 0 !important;
  cursor: pointer;
}

.c5 {
  box-sizing: border-box;
  font-family: -apple-system,BlinkMacSystemFont,"Segoe UI","Roboto","Oxygen", "Ubuntu","Cantarell","Fira Sans","Droid Sans","Helvetica Neue", sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  margin-top: 10px;
  margin-bottom: 0px;
  margin-left: 0px;
  margin-right: 0px;
  padding-top: 0px;
  padding-bottom: 0px;
  padding-left: 0px;
  padding-right: 0px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: start;
  -webkit-justify-content: start;
  -ms-flex-pack: start;
  justify-content: start;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c4 {
  box-sizing: border-box;
  margin: 0;
  padding: 0;
  font-family: -apple-system,BlinkMacSystemFont,"Segoe UI","Roboto","Oxygen", "Ubuntu","Cantarell","Fira Sans","Droid Sans","Helvetica Neue", sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  width: 100%;
  height: 32px;
  border: 2px solid #e5e7eb;
  border-radius: 5px;
  padding-left: 10px;
}

.c6 {
  width: 100%;
}

.c2 {
  width: 100%;
}

.c0 {
  width: 100%;
}

<div
  className="c0"
  data-testid="test-loginform-id"
>
  <div
    className="c1"
  >
    <div
      className="c2"
      data-testid="loginform-textinput-login"
    >
      <span
        className="c3"
        data-testid="textinput-label"
      >
        Email
      </span>
      <input
        className="c4"
        data-testid="textinput-textfield"
        id="email"
        name="Email"
        onChange={[Function]}
        type="email"
      />
    </div>
  </div>
  <div
    className="c1"
  >
    <div
      className="c2"
      data-testid="loginform-textinput-password"
    >
      <span
        className="c3"
        data-testid="textinput-label"
      >
        Password
      </span>
      <input
        className="c4"
        data-testid="textinput-textfield"
        id="password"
        name="Password"
        onChange={[Function]}
        type="password"
      />
    </div>
  </div>
  <div
    className="c5"
  >
    <div
      className="c6"
      data-testid="loginform-CheckboxArea"
    >
      <input
        className="c7"
        data-testid="checkboxarea-checkbox"
        id="remember"
        name="remember"
        type="checkbox"
      />
      <span
        className="c3"
        data-testid="checkboxarea-label"
      >
        Remember Me?
      </span>
    </div>
  </div>
  <div
    className="c1"
  >
    <button
      className="c8"
      data-testid="login-button-id"
      disabled={false}
      onClick={[Function]}
      type="submit"
    >
      LOGIN
    </button>
  </div>
  <div
    className="c9"
  >
    <button
      className="c10"
      color="rgb(107 114 128)"
      data-testid="loginform-link"
      onClick={[Function]}
      type="button"
    >
      Forgot Password?
    </button>
  </div>
</div>
`;
